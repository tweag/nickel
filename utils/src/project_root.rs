use std::path::PathBuf;

/// Return a [`PathBuf`] pointing to the workspace root of Nickel repository. For example, use
/// ```rust
/// # use nickel_lang_utilities::project_root::project_root;
/// # use std::path::PathBuf;
/// let readme = project_root().join("README.md");
/// # assert_eq!(readme, PathBuf::from(&concat!(env!("CARGO_MANIFEST_DIR"), "/../README.md")))
/// ```
///
/// This is a workaround for the lack of a CARGO_WORKSPACE_DIR
/// environment variable, as suggested [here](https://github.com/rust-lang/cargo/issues/3946#issuecomment-1433384192).
/// A better workaround might be to set this in the `[env]` section
/// of `.cargo/config.toml`, as suggested[here](https://github.com/rust-lang/cargo/issues/3946#issuecomment-973132993)
/// but we currently don't check that file in, and I'm not sure
/// whether doing so is a good idea.
pub fn project_root() -> PathBuf {
    PathBuf::from(&concat!(env!("CARGO_MANIFEST_DIR"), "/.."))
}
