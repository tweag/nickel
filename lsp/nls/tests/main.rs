use nickel_lang_utils::project_root::project_root;
use test_generator::test_resources;

use lsp_harness::{TestFixture, TestHarness};

#[test_resources("lsp/nls/tests/inputs/*.ncl")]
fn check_snapshots(path: &str) {
    let _ = env_logger::try_init();

    let full_path = project_root().join(path);

    let contents = std::fs::read_to_string(full_path).unwrap();
    let fixture = TestFixture::parse(&contents).unwrap();
    let mut harness = TestHarness::new();

    harness.prepare_files(&fixture);
    for req in fixture.reqs {
        harness.request_dyn(req);
    }

    harness.drain_diagnostics(fixture.expected_diags.iter().cloned());
    let output = String::from_utf8(harness.out).unwrap();

    insta::assert_snapshot!(path, output);
}

#[test]
fn refresh_missing_imports() {
    let _ = env_logger::try_init();
    let mut harness = TestHarness::new();

    let url = |s: &str| lsp_types::Url::from_file_path(s).unwrap();
    harness.send_file(url("/test.ncl"), "import \"dep.ncl\"");
    let diags = harness.wait_for_diagnostics().diagnostics;
    assert_eq!(2, diags.len());
    assert!(diags[0].message.contains("import of dep.ncl failed"));

    // We expect another copy of the diagnostics (coming from background eval).
    let diags = harness.wait_for_diagnostics().diagnostics;
    assert_eq!(2, diags.len());
    assert!(diags[0].message.contains("import of dep.ncl failed"));

    // Now provide the import.
    harness.send_file(url("/dep.ncl"), "42");

    // Check that we get back clean diagnostics for both files.
    // (LSP doesn't define the order, but we happen to know it)
    // Loop because we can get back the diagnostics twice from each
    // file (once from synchronous typechecking, once from eval in the background).
    loop {
        let diags = harness.wait_for_diagnostics();
        assert!(diags.diagnostics.is_empty());
        if diags.uri.path() == "/dep.ncl" {
            break;
        }
    }

    loop {
        let diags = harness.wait_for_diagnostics();
        assert!(diags.diagnostics.is_empty());
        if diags.uri.path() == "/test.ncl" {
            break;
        }
    }
}
