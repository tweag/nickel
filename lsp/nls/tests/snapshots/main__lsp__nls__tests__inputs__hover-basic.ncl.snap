---
source: lsp/nls/tests/main.rs
expression: output
---
<6:0-6:6>[```nickel
Dyn
```]
<6:0-6:10>[```nickel
Dyn
```, middle]
<6:0-6:14>[```nickel
Dyn
```, ```nickel
Number
```, innermost]

